x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -7,39 +7,47 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	common "github.com/dymensionxyz/dymension/v3/x/common/types"
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
 )
 
-// Called every block to finalize states that their dispute period over.
+// FinalizeQueue is called every block to finalize states when their dispute period over.
 func (k Keeper) FinalizeQueue(ctx sdk.Context) error {
 	if uint64(ctx.BlockHeight()) < k.DisputePeriodInBlocks(ctx) {
 		return nil
 	}
 	// check to see if there are pending  states to be finalized
 	finalizationHeight := uint64(ctx.BlockHeight() - int64(k.DisputePeriodInBlocks(ctx)))
 	pendingFinalizationQueue := k.GetAllFinalizationQueueUntilHeight(ctx, finalizationHeight)
+
+	return osmoutils.ApplyFuncIfNoError(ctx,
+		// we trap at this granularity because we want to avoid iterating inside the
+		// wrapped function (above), and we want the function to be atomic, so
+		// we cannot trap inside finalizePending
+		func(ctx sdk.Context) error {
+			k.finalizePending(ctx, pendingFinalizationQueue)
+			return nil
+		})
+}
+
+func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) {
```

Comment:
why u panic in this func instead of returning err?
it's much cleaner. u use cacheCtx on error also, don't have to panic

Reply:
I have changed the 
`keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)`
to be an error

But I'll keep the panic for the other one

I think it's a good idea to be disciplined about broken invariants, when they break, we should panic, because it means we have a logic bug and can't recover

Does it make sense?